### DISCUSSION TITLE
Java/android/hello

### DISCUSSION THREAD

**Discussion by ZiRaymondal** [zeroc-ice/ice-demos#134] [Q&A] [answered]
Hello ! 

I'm not able to connect https://github.com/zeroc-ice/ice-demos/blob/3.7/java/android/hello/src/main/java/com/zeroc/hello/HelloApp.java with the hello cpp11 demo. I tried to change tcp port on 10004 on the android demo to match server.config but it still not working.

It works with Ice Discovery tho...

**Comment by pepone** [comment]
are you running the android demo from the simulator, what IP address is the server listening on?

**Comment by ZiRaymondal** [comment]
Ok I fixed it : I need to setup the tcp port to 45049 ( line 392 of the last link), I found this port number with the start message of the cpp client ( endpoint line, first port i've seen lol ). Can you explain me why plz ? I'm kinda lost ( I'm not working directly on the demo but I'm doing a Univ project which starts from it ). Thanks for your reactivity btw

I'm running the demo on my current cellphone, the server seems to listen on 192.168.1.5:10004 ( that's what I understand from the serverconfig file of the cpp11 IceDiscovery/hello server & my ifconfig command ofc )

**Comment by pepone** [comment]
By default the server listens on   tcp 10000,  ssl 10001, ws 10002, and wss 10003 
https://github.com/zeroc-ice/ice-demos/blob/7a91b5461d8358753e766bd1b12cd505d1f093f1/cpp11/Ice/hello/config.server#L11

port 10004 is used for Admin object if enabled
https://github.com/zeroc-ice/ice-demos/blob/7a91b5461d8358753e766bd1b12cd505d1f093f1/cpp11/Ice/hello/config.server#L71

The simpler to check what address the server listens on, is to enable network tracing in the config https://github.com/zeroc-ice/ice-demos/blob/7a91b5461d8358753e766bd1b12cd505d1f093f1/cpp11/Ice/hello/config.server#L39

**Comment by ZiRaymondal** [comment]
[config.txt](https://github.com/zeroc-ice/ice-demos/files/6111360/config.txt)

I've started from the IceDiscovery/hello server not /Ice/hello, so I have this config.file ( not edited ). It doesn't work by default with the  tcp port set at 10000.

So i've just edited the admin port by mistake, but it I don't understand why I need to change tcp port from 10000 to the magic number 45049 on my android client.

Android Client line 39X  : is not working with the IceDiscovery/hello server
`        else
            s = "hello:tcp -h " + _host + " -p 10000:ssl -h " + _host + " -p 10001:udp -h " + _host  + " -p 10000";`
But it works with  tcp set to 45049

**Comment by pepone** [comment]
This demo works with the 2 servers to show different use cases:

* Ice/hello server uses a fixed address and port 10000, you must use this server when specifying the host address, in this case, the client application must know the address of the server

* IceDiscovery/hello this server doesn't use a fixed address and the client relies on the IceDiscovery location plug-in to find the server. In this case, you should select " Ice Discovery " to first discover the address of the server

see https://doc.zeroc.com/ice/3.7/ice-plugins/icediscovery


This is mention in the Android hello README https://github.com/zeroc-ice/ice-demos/tree/3.7/java/android/hello

**Reply by ZiRaymondal** [reply]
I have just a last question which importunates me xD : What are the cases where we would not use IceDiscovery ? It seems to avoid a lot of pain

**Reply by ZiRaymondal** [reply]
"IceDiscovery and IceGrid both provide a location service but it helps to understand their differences when deciding which one to use in an application. Use IceDiscovery when your application needs a lightweight, transient location service. IceGrid's location service is backed by a persistent database and represents just one of the features that IceGrid offers, along with remote administration, on-demand server activation, and many others. If you need a location service but aren't yet ready to dive into IceGrid, start out using IceDiscovery; migrating to IceGrid later won't be difficult." OK MB

THX AGAIN

**Reply by pepone** [reply]
Keep in mind that IceDiscovery relies on  UDP multicast to discover servers, it works well if the servers and clients run in the same local network.

**Reply by ZiRaymondal** [reply]
> Keep in mind that IceDiscovery relies on UDP multicast to discover servers, it works well if the servers and clients run in the same local network.

I was guessing that Discovery was made for connecting on the same local network but that was a pure feeling.

Thank you again for all the answerrrs !!

**Comment by ZiRaymondal** [comment]
Yep so ok I've missed that so far x'). Is there a way to print the Endpoint port once the Discovery  server launched ? 
I saw the client displays it but can't find how to do so server side.

Anyway thank u very much my IceLord for your time !

**Comment by pepone** [comment, marked as answer]
If you set `Ice.Trace.Network=3` in the server config file, you should see the address the server binds

https://doc.zeroc.com/ice/3.7/property-reference/ice-trace#id-.Ice.Trace.*v3.7-Ice.Trace.Network

**Comment by ZiRaymondal** [comment]
Thx very muchhhhh mann, you are so nice, have a nice day !
